ssl/statem/extensions_srvr.c

```diff
@@ -0,0 +1,934 @@
+        return 0;
+    }
+
+    if (PACKET_remaining(&exts) > 0) {
+        const unsigned char *ext_data = PACKET_data(&exts);
+
+        sk_X509_EXTENSION_pop_free(s->tlsext_ocsp_exts,
+                                   X509_EXTENSION_free);
+        s->tlsext_ocsp_exts =
+            d2i_X509_EXTENSIONS(NULL, &ext_data, (int)PACKET_remaining(&exts));
+        if (s->tlsext_ocsp_exts == NULL || ext_data != PACKET_end(&exts)) {
+            *al = SSL_AD_DECODE_ERROR;
+            return 0;
+        }
+    }
+
+    return 1;
+}
+#endif
+
+#ifndef OPENSSL_NO_NEXTPROTONEG
+int tls_parse_ctos_npn(SSL *s, PACKET *pkt, int *al)
+{
+    /*
+     * We shouldn't accept this extension on a
+     * renegotiation.
+     *
+     * s->new_session will be set on renegotiation, but we
+     * probably shouldn't rely that it couldn't be set on
+     * the initial renegotiation too in certain cases (when
+     * there's some other reason to disallow resuming an
+     * earlier session -- the current code won't be doing
+     * anything like that, but this might change).
+     *
+     * A valid sign that there's been a previous handshake
+     * in this connection is if s->s3->tmp.finish_md_len >
+     * 0.  (We are talking about a check that will happen
+     * in the Hello protocol round, well before a new
+     * Finished message could have been computed.)
+     */
+    if (s->s3->tmp.finish_md_len == 0)
+        s->s3->next_proto_neg_seen = 1;
+
+    return 1;
+}
+#endif
+
+/*
+ * Save the ALPN extension in a ClientHello.|pkt| holds the contents of the ALPN
+ * extension, not including type and length. |al| is a pointer to the alert
+ * value to send in the event of a failure. Returns: 1 on success, 0 on error.
+ */
+int tls_parse_ctos_alpn(SSL *s, PACKET *pkt, int *al)
+{
+    PACKET protocol_list, save_protocol_list, protocol;
+
+    if (s->s3->tmp.finish_md_len != 0)
+        return 1;
+
+    if (!PACKET_as_length_prefixed_2(pkt, &protocol_list)
+        || PACKET_remaining(&protocol_list) < 2) {
+        *al = SSL_AD_DECODE_ERROR;
+        return 0;
+    }
+
+    save_protocol_list = protocol_list;
+    do {
+        /* Protocol names can't be empty. */
+        if (!PACKET_get_length_prefixed_1(&protocol_list, &protocol)
+                || PACKET_remaining(&protocol) == 0) {
+            *al = SSL_AD_DECODE_ERROR;
+            return 0;
+        }
+    } while (PACKET_remaining(&protocol_list) != 0);
+
+    if (!PACKET_memdup(&save_protocol_list,
+                       &s->s3->alpn_proposed, &s->s3->alpn_proposed_len)) {
+        *al = TLS1_AD_INTERNAL_ERROR;
+        return 0;
+    }
+
+    return 1;
+}
+
+#ifndef OPENSSL_NO_SRTP
+int tls_parse_ctos_use_srtp(SSL *s, PACKET *pkt, int *al)
+{
+    STACK_OF(SRTP_PROTECTION_PROFILE) *srvr;
+    unsigned int ct, mki_len, id;
+    int i, srtp_pref;
+    PACKET subpkt;
+
+    /* Ignore this if we have no SRTP profiles */
+    if (SSL_get_srtp_profiles(s) == NULL)
+        return 1;
```

Comment:
We might be more friendly to the ecosystem if we moved this check after the parity check that happens next.